osd: fix cleanup job disk shredding

puskunalis opened the pull request:
Fixes quick disk shredding by using `dd` to shred data at additional offsets where ceph metadata is duplicated. For full shred, the `shred` utility remains in use.

<!-- Thank you for contributing to Rook! -->

<!-- STEPS TO FOLLOW:
  1. Add a description of the changes (frequently the same as the commit description)
  2. Enter the issue number next to "Resolves #" below (if there is no tracking issue resolved, **remove that section**)
  3. Review our Contributing documentation at https://rook.io/docs/rook/latest/Contributing/development-flow/
  4. Follow the steps in the checklist below, starting with the **Commit Message Formatting**.
-->

**Issue resolved by this Pull Request:**
Resolves #15546

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.

sp98 commented on .github/workflows/canary-integration-test.yml:
```diff
@@ -770,6 +770,121 @@ jobs:
+      - name: check-ownerreferences
+        run: tests/scripts/github-action-helper.sh check_ownerreferences
+
+      - name: teardown cluster with cleanup policy
```
Instead of a new CI to test the cleanup, I would suggest integrating this test in one of the existing CIs.

sp98 commented:
is this ready for review @puskunalis ?

puskunalis commented:
Not yet @sp98 I will mark it as ready and re-request your review once it is

puskunalis commented:
@sp98 it’s ready for review now, but I would need some guidance on what additional tests to add - encryption-pvc cleanup has issues, others have issues starting the cleanup

sp98 commented:
> @sp98 it’s ready for review now, but I would need some guidance on what additional tests to add - encryption-pvc cleanup has issues, others have issues starting the cleanup

what issues are you facing with the encryption-pvc ci?  
Can you also add the cleanup job logs from your local testing. Thanks.

puskunalis commented:
@sp98 everything works fine locally, but as encryption-pvc sets up different volumes (/dev/mapper/...) I think it is cleared at the wrong offsets. It only works if /dev/sd* are cleaned up. This is not really related to my change (if I understand correctly), it's just that this hasn't been tested before

puskunalis commented:
@travisn I will try to have a look into this over the next week

puskunalis commented:
@sp98 could you please have a look why encryption-pvc is failing here for an encrypted PVC drive? I'm suspecting something with the test configuration, as it is not marked as a PVC OSD: https://github.com/rook/rook/actions/runs/14979187015/job/42078958350

Also, is there a reason why `shred` was used instead of `ceph-volume lvm zap` for disk cleanup? It looks as `ceph-volume lvm zap` might be the cleaner option.

sp98 commented:
> Also, is there a reason why `shred` was used instead of `ceph-volume lvm zap` for disk cleanup? It looks as `ceph-volume lvm zap` might be the cleaner option.

I'm not sure of the exact reason. It was written a few years ago. If you see, there are two code paths here. One for `lvm` mode OSD and other for `raw` mode OSDs. `lvm` mode OSDs use `ceph-volume lvm zap` and `raw` mode OSDs use `shred`. 
But like you said, `ceph-volume lvm zap` can be used for both the modes. The command name is confusing and one would think it only works for lvm mode, which is not true.  I also agree on just using `ceph-volume lvm zap` for all the disk cleanups.  Few thoughts around this: 

- This would mean less complexity in the code base. 
- We will have to deprecate the [`shred` utility (`sanitizeDisks`)](https://rook.io/docs/rook/latest/CRDs/Cluster/ceph-cluster-crd/#cleanup-policy) based API in the cephCluster.  I think that should be ok because cluster cleanup is not a very common scenario and users won't mind just cleaning up the whole disk during cleanup. 
- We will lose the advantages of `shred` utility which we are anyways going if we implement `dd` for cleanup of disks. 
- There [was an issue](https://github.com/ceph/ceph/pull/62177) with `ceph-volume lvm zap` and partitioned disks. It got fixed recently. 

@travisn @puskunalis @BlaineEXE Any further thoughts on whether we just use `ceph-volume lvm zap` in the cleanup job

puskunalis commented:
We could keep using shred for full disk shredding, but ceph-volume lvm zap for quick shred (otherwise we still need to use dd instead of shred)

puskunalis commented:
Is ceph-volume not already in the Room image? In this PR I currently set it to run that command and the tool is present, unless the CI setup is different somehow.

puskunalis commented on pkg/daemon/ceph/cleanup/disk.go:
```diff
@@ -174,25 +191,107 @@ func (s *DiskSanitizer) buildShredArgs(disk string) []string {
 		}
 
-		output, err := s.context.Executor.ExecuteCommandWithCombinedOutput(shredUtility, s.buildShredArgs(device)...)
+		// output, err := s.context.Executor.ExecuteCommandWithCombinedOutput("ceph-volume", "lvm", "zap", "--destroy", "--osd-id", strconv.Itoa(osdInfo.ID))
```
Sounds good, but I would need some guidance on encrypted PVC drives as the test is currently failing, it is reporting PVC name as empty, so maybe something specific with the test segtup?

Also do we need to execute this command or would it make sense to reuse the DestroyOSD function?

sp98 commented:
> Hey @sp98, @travisn, thanks for your patience, this is ready for review now.
> 
> Some notes:
> 
> * `ceph-volume lvm zap` doesn't work on `/dev/mapper/set1-data-0abc1-block-dmcrypt` devices, so we need to find the backing device. The method in the `DestroyOSD` function using `/mnt` doesn't work as PVC volumes aren't mounted into the cleanup pod. The function also fails to detect them as PVCs as the necessary environment variables are missing somewhere.
> * `cryptsetup` is used to discover the backing device for the `/dev/mapper/set1-data-0abc1-block-dmcrypt` device. I had no luck with `lsblk`, even after changing some flags. Let me know if this can be improved.

I think that should be ok.
OSD prepare pod with `mnt/` path mounted uses [EncryptionDMName](https://github.com/rook/rook/blob/a44940d924bc4231fe4bec4bcc01752e48f7d157/pkg/daemon/ceph/osd/remove.go#L271C20-L271C36) function to get the  backing device using lsblk and then clean up. 

> * This way, the quick+random shred method is no longer random and is now the same as quick+zero - might need to document this somewhere? I think there is no use case for randomly shredding only tiny parts of the data on the disk.
> * The cleanup flow was only added to `pvc` and `encryption-pvc` tests - are there any other tests it would make sense to be added to?
> * I assume the two failing integration tests are simply flaky?
Known issue with Keystone Auth. Not related to this PR. 

> * I might have not considered encryption on non-PVC OSDs, any pointers here?

puskunalis commented on pkg/daemon/ceph/cleanup/disk.go:
```diff
@@ -174,32 +174,62 @@ func (s *DiskSanitizer) buildShredArgs(disk string) []string {
 
+	// If the device is encrypted, get the real path and remove the dm device
+	if osdInfo.Encrypted {
+		realPath, err := osd.GetBackingDeviceForEncryptedBlock(s.context, osdInfo.BlockPath)
```
Yes, with `PopulateDeviceInfo` it returns something like `/dev/dm-1` as `RealPath` (`NAME`), but `PKNAME` is empty even when `--no-deps` is removed from the `PopulateDeviceInfo` `lsblk` call.

sp98 commented on .github/workflows/canary-integration-test.yml:
```diff
@@ -752,6 +752,44 @@ jobs:
+          kubectl -n rook-ceph patch cephcluster rook-ceph --type merge -p '{"spec":{"cleanupPolicy":{"confirmation":"yes-really-destroy-data"}}}'
+          kubectl -n rook-ceph delete cephcluster rook-ceph
+          kubectl -n rook-ceph logs deploy/rook-ceph-operator
+          tests/scripts/github-action-helper.sh wait_for_cleanup_pod
```
I think it was fixed along with https://github.com/ceph/ceph/pull/62177

puskunalis commented:
Moved the cluster teardown logic into a new function and added some docs about quick+random cleanup being the same as quick+zero. Let me know if you need me to change anything else

sp98 commented on pkg/daemon/ceph/cleanup/disk.go:
```diff
@@ -174,32 +174,62 @@ func (s *DiskSanitizer) buildShredArgs(disk string) []string {
+	if osdInfo.Encrypted {
+		realPath, err := osd.GetBackingDeviceForEncryptedBlock(s.context, osdInfo.BlockPath)
+		if err != nil {
+			logger.Errorf("failed to get backing device for encrypted block %q. %v", osdInfo.BlockPath, err)
```
`realpath` won't be available if there any error.  Should we continue after the error?

puskunalis commented:
Hey @sp98 @travisn anything else needed from my side? If you'd prefer me to, I can hand over this PR to you.

sp98 reviewed (commented):
@puskunalis Just a couple of questions?  Other than that the PR looks good. Approving.

puskunalis commented on .github/workflows/canary-integration-test.yml:
```diff
@@ -752,6 +752,42 @@ jobs:
         run: tests/scripts/github-action-helper.sh check_ownerreferences
 
+      - name: teardown cluster with cleanup policy
+        run: |
```
Created https://github.com/rook/rook/pull/16119
